### ISSUE TITLE
code=TRUE does not substitute in other variables.

### ISSUE THREAD

**Issue by nfultz** [DeclareDesign/designs#31]
simple_two_arm_designer(N=200, code=TRUE)   does not return a N=200 design with the 200 substituted in, but instead returns a block of code where N is still a free variable. code=TRUE essentially ignores all the other arguments

Having a flag that changes the return type is usually weak design - see eg https://martinfowler.com/bliki/FlagArgument.html - you may be better off just having users do `get_design_code(my_design)`, instead of copying the code=TRUE boilerplate into every single design.

**Comment by jaspercooper** [comment]
Thanks for raising this substitution issue, we'd really like to solve it. 

Let's say you have a function:
```r
test <- function(m = 50,N = 100){
  stopifnot(m < N)
  {{{
    mean(N - m)
  }}}
}
```
What we'd like is a get_design_code function that would return as a string something like the output of this for (A) the defaults:
```r
substitute(
  mean(N - m),
  list(N = 100, m = 50)
)
```
Then if the user (B) supplied `N = N`, something that returned a string similar to the output of:
```r
substitute(
  mean(N - m),
  list(N = as.name("N"), m = 50)
)
```
or, if the user (C) supplied `N = 2, m = 1`, a string similar to the output of:
```r
substitute(
  mean(N - m),
  list(N = 2, m = 1)
)
```
Would love to hear your thoughts on how to make `get_design_code` do this. Use case B is important for declaring designs that have only one modifiable argument, for example. A and C are important for the reason you mention, and would have the added advantage of freeing us from the verbosely titled `designer_default_args_text`. They'd also mean the shiny app code could be copied and pasted directly, irrespective of where you were in the parameter space (at the moment we are doing this through a crazier version of `designer_default_args_text` that I'm not particularly fond of).

Any thoughts on how to achieve A B and C in `get_design_code`?

**Comment by macartan** [comment]
Thanks JC; I had jimmied up the attached; advantage here is that people can
1. provide no defaults and defaults are provided all as modifiable arguments
2. provide different values, keeping all as modifiable arguments
3. fix some values which get substituted in and no longer appear as arguments in the code printout, or as modifiable arguments

**Comment by nfultz** [comment]
Also check if it works with comments, iirc that's why getsrcref was used
originally.

**Comment by nfultz** [comment]
Also I think substitute() will break designs that use the magic N in
fabricatr eg for hierarchical data.

**Comment by macartan** [comment]
Do see another strategy to insert values of some fixed  arguments into code
print outs?

**Comment by macartan** [comment]
[Here](https://github.com/DeclareDesign/designs/files/1957929/m_arm_designer.pdf) is a go at a designer where some expressions are pasted together but we want the code display to be clean.  This uses the same substitute approach:
* Note that comments are handled here, but other formatting is not good yet
* The estimand declaration is not working properly because of quotes 

Would be good to see solutions either using substitute or something else that deals with the paste issues and optionally lets people fix values of some parameters while leaving others modifiable.

code below:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls())
library(DeclareDesign)
library(knitr)
library(DesignLibrary)
```

Trying to get nice code print out when complex expressions are made in text. Issues:

* Not getting estimands declared properly becuase of use of ""; no luck yet with cat or noquote etc
* Need some wrapping on the text output


```{r}
cleanup_code <- function(txt)   {
    indentation <- strsplit(as.character(txt)[1], "")[[1]]
    indentation <- indentation[cumprod(indentation == " ") == 1]
    indentation <- paste0("^", paste(indentation, collapse=""))
    sub(indentation, "", txt)[-1]
    }

designer <- function(
    N = 30, n_arms = 3, 
    means = rep(0, n_arms),
    conds = paste0("T", 1:n_arms),
    sd = 1,   
    f_Y = formula(paste0(
              "Y ~ ", paste0(means, " * (Z == 'T", 1:n_arms, "')", collapse = " + "),  " + u")
              ),
    mand =  paste("baseline = Y_Z_T1,",
             paste0("T", 2:n_arms, "T1 = Y_Z_T", 2:n_arms, "- Y_Z_1", collapse = ", ")),
    fixed = list(f_Y = f_Y, conds = conds, mand = mand, n_arms = n_arms),
    code = FALSE){
  
            
  design_code <-  substitute( 
    {
    "# M: Model ##############################################################"
    pop <- declare_population(N = N, u = sd*rnorm(N))
    pos <- declare_potential_outcomes(formula = f_Y, conditions = conds)


    "# I: Inquiry ############################################################"
    esm <- declare_estimand(mand, coefficients = TRUE)

    "# D: Data ###############################################################"
    asn <- declare_assignment(num_arms = n_arms)
    
    "# A: Answer #############################################################"

    est <- declare_estimator(Y ~ Z, model = lm_robust, coefficients = TRUE)
    
    declare_design(pop, pos, asn, declare_reveal(), est)
    
     }, fixed)
 
  if(code)  out <- cleanup_code(design_code)
  
  if(!code) out <- eval(design_code)
  return(out)
  }


```

# Example

Here is the code for a design (displayed as code using chunk option):

```{r, eval = FALSE}
designer(N = 30, n_arms = 4,  code = TRUE)
```

`{r, code = designer(N = 30, n_arms = 4, code = TRUE), eval = FALSE, tidy = TRUE}`



```{r, comment = "", error = TRUE}


kable(head(draw_data(designer(n_arms = 5, means = 1:5))), digits = 2)

design <- designer(n_arms = 5)
design[[5]](draw_data(design))

diagnose_design(design)
```

**Comment by macartan** [comment]
Last desideratum and concern with substitute approach; *it would be good to be able to have functions as arguments*:

eg the below works fine 
```{r}
designer <- function(N = 10, f = rnorm){
  
  pop <- declare_population(N = N, X = f(N))
  declare_design(pop)

  }

draw_data(designer(N = 5))
draw_data(designer(N = 5, f = runif))
```
But substitute approach breaks

**Comment by nfultz** [comment]
You might be able to manually splice in quosures by using rlang throughout.

**Comment by jaspercooper** [comment]
Thanks for this. This is very similar to the way it was originally set up and also the vignettes do something similar using `designer_default_args_text` to assign formals. 

It's also my preferred option and something @macartan and I have gone back and forth on quite a bit. 

I'm sympathetic to his view, which is that we shouldn't impose on users any particular methodology for returning code (i.e., by using {{{), but simply the constraint that the designer function needs to return code somehow whenever code = TRUE. 

The principal advantage to that approach as I understand it is that some designs require quite complex code. In particular those whose dimensionality or particular dgp functions are not known in advance (multi-arm, for example). Whether your solution is to rlang or eval(parse(text = paste(stuff))), the unsubstituted code is a bit ugly and hard to interpret for such designers.

**Comment by macartan** [comment]
Thanks ; yes, this is looking similar to what we are currently using I
think (except code is attribute rather than option) and I am not sure this
addresses the two challenges that are worrying us:

1. how to substitute values into the printed code; so that
get_design_code(my_two_arm_designer(N = 50))

would print out
##
## # M: Model
## pop <- declare_population(
## N = 50,
## Z0 = rnorm

2. How to handle the awkward pasting stuff needed for complex generated
statements; as as in the multiarm examples (but sure to be more complex
with more complex examples)

**Comment by macartan** [comment]
PS -- I do in principle like the idea of the code being an attribute of a
design; stan objects similarly include the code and much more
A challenge I see though is what happens when you modify the design, eg
design2 <- replace_step(design, 1, pop2), will have the wrong code unless
we update attributes also

**Comment by jaspercooper** [comment]
At present the code is not an attribute of the designer(). We could maybe add it as an attribute recursively but we still have the issue of finding a way we all like to extract the function's body (via the get_design_code() + {{{ method or via code = TRUE method).

**Comment by graemeblair** [comment]
My sense is trying to substitute in the code is simply too hard a problem to solve -- or one we'd have spend quite a bit of time working out. I don't immediately see the payoff -- having

N <- 50
function <- rnorm

seems quite clear to me. 

I agree if you modify the design it should modify the code. I think we may need to think beyond the designers about how to print and include code in objects. There may not be a single solution that works for both.

**Comment by acoppock** [comment]
So `code = TRUE` violates a core principle of our packages, which is that functions should always return objects of the same type.  Designers have to consistently return a design object.

The attribute approach solves the problem in a clear way; agree that if modified, the code attribute wouldn't also update. I think this unfortunate consequence is worth the tremendous gain in terms of things not breaking.

**Comment by macartan** [comment]
I am happy as  a solution to have a designer return a list with design and
code which is small edit and keeps things together; a bit more awkward to
have design <- designer(N = 10)[[1]] but if you feel this principle is
important that's a cost to pay

Having a design carry code as an attribute that is not consistent with it
sounds to me orders of magnitude more dangerous than having a code option

But eyes on the prize: the big issues are being able to substitute in --
not just N but nasty pasted strings -- and maintaining flexibility; lets
talk through tomorrow

**Comment by acoppock** [comment]
yep, the real solution here is getting something excecutable from `get_code(design)`, but I honestly don't know if that is possible. 

With the designers, it's possible because we're just grabbing the literal text in between the `{{{` and `}}}`, which is a nice solution to a hard problem -- have intention about the code that will be generated on the front end, rather than trying to infer it from a design.

Maybe we delete the code attribute if it's modified?  In any case, I think designers should return designs, not lists. Our package is built towards building this one cool object type that you can then do things with, so I think the list solution isn't right either.
